Add "low-rank" variational families

Red-Portal opened the pull request:
This PR adds low-rank variational families which cannot be simply represented as a location-scale family (the reparameterization path has to be modified.)

* [x] add `MvLocationLowRankScale`
* [x]  implement `rand`
* [ ] implement `logpdf`
* [ ] implement `entropy`
* [x] implement `mean`, `var`, and `cov`
* [ ] add bunch of tests

The tricky part would be `logpdf` and `entropy` since, to be done efficiently, will have to involve low-rank Cholesky updates. Given that low-rank Cholesky updates are niche, I am not sure whether their AD is up to the task.

Red-Portal commented:
Okay the tests didn't pass because the latest Enzyme patch somehow broke some stuff, but all tests passed locally.

mhauru reviewed (changes requested):
Hi @Red-Portal! I was actually just yesterday thinking about something like `MvLocationScale`, I hadn't realised it existed already. This low-rank version seems cool too.

I put a bunch of questions and proposals I had in local comments. It's all about code stuff, I don't have much expertise on the theory side here.

More broadly, and probably not to be addressed in this PR, but is there a reason to keep `MvLocationScale` and `MvLocationScaleLowRank` in AdvancedVI, rather than somewhere more centrally in TuringLang so that one could use them more broadly with Turing.jl, or maybe even in Distributions.jl?

I suspect the main reason @yebai tagged me as a reviewer though is the Enzyme failure. I'll look into it.

mhauru commented:
The Enzyme issue is this: The function `A = randn(D, D); f(λ′) = λ′'*A*λ′ / 2` in `test/interface/ad.jl` creates a closure, and then we call AD on it as
```
    _, y = Enzyme.autodiff(
        Enzyme.ReverseWithPrimal,
        f,
        Enzyme.Active,
        Enzyme.Duplicated(θ, ∇θ),
    )
```
Since `f` is a closure Enzyme is unable to figure out whether it might be mutated or contain differentiable data. (Note that a callable is an object like any other, and thus all sorts of things could be stored in it.) This can be fixed by changing the call to
```
    _, y = Enzyme.autodiff(
        Enzyme.ReverseWithPrimal,
        Enzyme.Const(f),  # Only line that has changed
        Enzyme.Active,
        Enzyme.Duplicated(θ, ∇θ),
    )
```
which is just explicitly telling Enzyme that you, the caller, guarantee that `f` does not contain anything that needs differentiating. That's probably the right thing to do, though I don't know if there could be some situation where the assumption would not be true, in which case it would of course be very bad to mark `f` as a constant if it's not.

Red-Portal commented on docs/src/elbo/families.md:
````diff
@@ -0,0 +1,122 @@
+z \sim  q_{\lambda} \qquad\Leftrightarrow\qquad
+z \stackrel{d}{=} C u + m;\quad u \sim \varphi
+```
+where ``C`` is the *scale*, ``m`` is the location, and ``\varphi`` is the *base distribution*.
````
Yup it works!

Red-Portal commented on docs/src/elbo/families.md:
````diff
@@ -0,0 +1,122 @@
+q = MeanFieldGaussian(μ, L)
+```
+
+### Sudent-$$t$$ Variational Families
````
Yup it works!

Red-Portal commented on test/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,148 @@
+            q_trans = if isnothing(bijector) 
+                q
+            else
+                Bijectors.TransformedDistribution(q, identity)
```
Thanks for spotting this!

Red-Portal commented on test/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,148 @@
+
+    @testset "diagonal positive definite projection" begin
+        @testset "$(realtype) $(bijector)" for
+            realtype    = [Float32,     Float64],
+            bijector    = [nothing,    :identity]
```
You're right. Just fixed this now.

Red-Portal commented on test/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,148 @@
+
+@testset "interface LocationScaleLowRank" begin
+    @testset "$(basedist) rank=$(rank) $(realtype)" for
+        basedist = [:gaussian],
```
I think something with skew would work, but I think in that case it would only make sense to also learn the skew, which is not quite supported for `LocationScale`.  Maybe `LocationScale` could be extended to support skew. If that is not the case, we can always make a new type that would work for that.

But I do know that there currently is a bug where `mean` and `var` is not correct unless `basedist` is standardized. I'll have to fix this in a different PR and maybe add some non-Gaussian test cases then.

Red-Portal commented on src/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,162 @@
+    U::Matrix{T};
+    scale_eps::T = sqrt(eps(T))
+) where {T <: Real}
+    @assert minimum(D) ≥ sqrt(scale_eps) "Initial scale is too small (smallest diagonal scale value is $(minimum(D)). This might result in unstable optimization behavior."
```
Oops that's a mistake because I added `scale_eps` later. Nice catch.

Red-Portal commented on src/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,162 @@
+
+function StatsBase.entropy(q::MvLocationScaleLowRank)
+    @unpack location, scale_diag, scale_factors, dist = q
+    n_dims  = length(location)
```
Ran the formatter for this one!

Red-Portal commented on src/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,162 @@
+
+function Distributions.logpdf(q::MvLocationScaleLowRank, z::AbstractVector{<:Real})
+    @unpack location, scale_diag, scale_factors, dist = q
+    #
```
I think I will keep it there for future reference.

Red-Portal commented on src/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,162 @@
+    return x .+= location
+end
+
+Distributions.mean(q::MvLocationScaleLowRank) = q.location
```
That's actually something that we need to enforce. I'll deal with this in a separate PR.

Red-Portal commented on src/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,162 @@
+
+Distributions.mean(q::MvLocationScaleLowRank) = q.location
+
+function Distributions.var(q::MvLocationScaleLowRank)  
```
Same with `mean`. This is actually a bug and will need to fix it later.

Red-Portal commented on src/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,162 @@
+- `U::Matrix{T}`: Low-rank factors of the scale, where `size(U,2)` is the rank.
+
+# Keyword Arguments
+- `check_args`: Check the conditioning of the initial scale (default: `true`).
```
Nice catch. Thanks!

Red-Portal commented on src/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,162 @@
+    dist         ::ContinuousDistribution;
+    scale_eps    ::T = sqrt(eps(T))
+) where {T <: Real}
+    @assert size(scale_factors,1) == length(scale_diag)
```
I was originally concerned about something, but I think it was unfounded. I'll move the checks to the general constructor. Thanks for the suggestion!

Red-Portal commented on src/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,162 @@
+
+Base.size(q::MvLocationScaleLowRank) = size(q.location)
+
+Base.eltype(::Type{<:MvLocationScaleLowRank{S, D, L}}) where {S, D, L} = eltype(D)
```
Nice catch. It appears `LocationScale` also had wrong type variables after the addition of `scale_eps`. But for some reason the tests didn't catch them for some reason.

Red-Portal commented on docs/src/elbo/families.md:
````diff
@@ -0,0 +1,145 @@
+q = MeanFieldGaussian(μ, L)
+```
+
+### Sudent-$$t$$ Variational Families
````
Nice catch thanks!

Red-Portal commented on docs/make.jl:
```diff
@@ -2,23 +2,23 @@
+        "ELBO Maximization" => [
+            "Overview" => "elbo/overview.md",
+            "Reparameterization Gradient Estimator" => "elbo/repgradelbo.md",
+            "Variational Families" => "elbo/families.md",
```
That's a good suggestion. Let me do it.

Red-Portal commented:
Hi @sunxd3! Thanks for the comments!  #75 was mentioned in a response to a reviewed comment by @mhauru . I'm planning to fix #75 after this PR gets merged.

mhauru reviewed (commented):
Looks good to me apart from the bug where LocationScaleLowRank assumes zero mean and unit variance. I know you said you'd fix that in a separate PR, but I'm a bit wary of committing to master code that has a known, significant bug in it. Could we make it so that the code errors out if the assumption of normalisation is violated? I can just imagine a situation where this gets into master, someone tries to use it before the other PR gets done, and it silently gives wrong results.

Red-Portal commented:
@mhauru Yeah, now that I think about it, I think there isn't much reason to solve it in a separate PRs. Let me address it today.

Red-Portal commented:
Adding the test cases for non-standardized base distributions is currently bottlenecked upstream by a type instability in Distributions.jl. I submitted a [PR](https://github.com/JuliaStats/Distributions.jl/pull/1885) and accompanying [issue](https://github.com/JuliaStats/Distributions.jl/issues/1884).  As the fix is mild, I expect this to be passed soon. Then I think this PR will be ready to go.

Red-Portal commented:
Hi @mhauru ! I fixed everything, added tests that regress #75 , and fixed the weird for loop indentations in the tests (though this means the formatter will complain about it). I also wanted to add some tests with a `StudentT` base distribution, but some type instability issues in `Distributions.jl` ([here](https://github.com/JuliaStats/Distributions.jl/issues/1884)) are going to need some time to get ironed out. So I gave up on that for now.

Red-Portal commented:
@mhauru Gentle nudge

mhauru commented on src/families/location_scale.jl:
```diff
@@ -113,16 +113,21 @@ function Distributions._rand!(
 function Distributions.cov(q::MvLocationScale)
     C = q.scale
-    return Hermitian(C * C')
+    σ2 = var(q.dist)
+    return σ2 * Hermitian(C * C')
```
I don't know the theory here well, but is there a reason why this involves `var(q.dist)` rather than `cov(q.dist)`? I could have imagined it being something like `C * cov(q.dist) * C'`, though that's just a not-very-educated guess.

Red-Portal replied:
Good point. I was thinking that `q.dist` was constrained to be a univariate distribution, which would make all of this valid, but seems like I have to use `ContinuousUnivariateDistribution` for that. Let me fix this later.

mhauru replied:
Oh I see, yeah, this makes sense for univariate. Is there a reason you want to restrict to `q.dist` being univariate? Just less of a headache to implement?

mhauru commented on src/families/location_scale_low_rank.jl:
```diff
@@ -111,18 +119,25 @@ function Distributions._rand!(
 function Distributions.cov(q::MvLocationScaleLowRank)
     @unpack scale_diag, scale_factors = q
-    return Diagonal(scale_diag .* scale_diag) + scale_factors * scale_factors'
+    σ2 = var(q.dist)
```
Same question here as above.

Red-Portal replied:
Same response as above. It is valid if `q.dist` is properly constrained to be a univariate distribution.

mhauru commented on test/runtests.jl:
```diff
@@ -50,6 +50,8 @@ if GROUP == "All" || GROUP == "Interface"
     include("interface/optimize.jl")
     include("interface/repgradelbo.jl")
+    include("interface/rules.jl")
+    include("interface/averaging.jl")
```
Are these an unrelated fix, where the files were there but not being included? Just wondering since these files seem to not have been edited in this PR. I'm happy to have an unrelated fix in this PR, just want to understand.

Red-Portal replied:
Not sure how this ended up here, but it seems like it went away now.

mhauru commented on test/families/location_scale.jl:
```diff
@@ -1,27 +1,33 @@
         elseif covtype == :meanfield && basedist == :gaussian
-            MeanFieldGaussian(μ, L)
+            MeanFieldGaussian(location, scale)
+        elseif covtype == :fullrank && basedist == :gaussian_nonstd
+            MvLocationScale(location, scale, Normal(realtype(3), realtype(3)))
+        elseif covtype == :meanfield && basedist == :gaussian_nonstd
+            MvLocationScale(location, scale, Normal(realtype(3), realtype(3)))
```
Would it make sense to test also a case where the base distribution is multivariate, with a non-trivial covariance matrix? Probably just something like `c = randn(n_dims, n_dims); MvNormal(randn(n_dims), c*c')`? Some code paths (e.g. the call to `Fill` in `mean`) might work differently depending on the dimension of the underlying distribution.

Red-Portal replied:
We probably never want the base distribution to be multivariate.

mhauru commented on src/families/location_scale.jl:
```diff
@@ -113,16 +113,21 @@ function Distributions._rand!(
-Distributions.mean(q::MvLocationScale) = q.location
+function Distributions.mean(q::MvLocationScale)
+    @unpack location, scale = q
+    return location + scale * Fill(mean(q.dist), length(location))
```
What does this do in a case where e.g. `location` is of dimension 3 and `q.dist` is of dimension 2? Or worse, have those two numbers be the other way around? Or does something (the constructor?) already earlier catch that?

Red-Portal replied:
Should be fixed by forcing `q.dist` to always be univariate.

mhauru reviewed (commented):
Thanks @Red-Portal, and sorry for being a bit slow to respond. I had a few more questions. I'll also need to build and read the new docs still, haven't done that yet.

mhauru commented:
Docs look good. I still wonder if these distributions would be useful more broadly than just within AdvancedVI.

Red-Portal commented:
@yebai The original formatter complaints were manual touches suggested by @mhauru because the formatted did a pretty ugly job in a few places. But I guess it will be hard to do manual formatting in the long run since people will just want to run the formatter.

mhauru commented on test/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,177 @@
+        @testset "statistics" begin
+            @testset "mean" begin
+                @test eltype(mean(q)) == realtype
+                @test mean(q) == mean(q_true)
```
```suggestion
                @test mean(q) ≈ mean(q_true)
```
This might fix the failing test.

mhauru commented on test/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,177 @@
+                @test cov(z_samples; dims=2) ≈ cov(q_true) rtol = realtype(1e-2)
+
+                z_sample_ref = rand(StableRNG(1), q)
+                @test z_sample_ref == rand(StableRNG(1), q)
```
```suggestion
                @test z_sample_ref ≈ rand(StableRNG(1), q)
```

mhauru commented on test/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,177 @@
+                @test cov(z_samples; dims=2) ≈ cov(q_true) rtol = realtype(1e-2)
+
+                samples_ref = rand(StableRNG(1), q, n_montecarlo)
+                @test samples_ref == rand(StableRNG(1), q, n_montecarlo)
```
```suggestion
                @test samples_ref ≈ rand(StableRNG(1), q, n_montecarlo)
```

mhauru commented on test/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,177 @@
+                end
+                z_samples = mapreduce(first, hcat, res)
+                z_samples_ret = mapreduce(last, hcat, res)
+                @test z_samples == z_samples_ret
```
```suggestion
                @test z_samples ≈ z_samples_ret
```

mhauru commented on test/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,177 @@
+
+                z_sample = Array{realtype}(undef, n_dims)
+                rand!(StableRNG(1), q, z_sample)
+                @test z_sample_ref == z_sample
```
```suggestion
                @test z_sample_ref ≈ z_sample
```

mhauru commented on test/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,177 @@
+            @testset "rand! AbstractMatrix" begin
+                z_samples = Array{realtype}(undef, n_dims, n_montecarlo)
+                z_samples_ret = rand!(q, z_samples)
+                @test z_samples == z_samples_ret
```
```suggestion
                @test z_samples ≈ z_samples_ret
```

mhauru commented on test/families/location_scale_low_rank.jl:
```diff
@@ -0,0 +1,177 @@
+
+                z_samples = Array{realtype}(undef, n_dims, n_montecarlo)
+                rand!(StableRNG(1), q, z_samples)
+                @test z_samples_ref == z_samples
```
```suggestion
                @test z_samples_ref ≈ z_samples
```

mhauru reviewed (commented):
A test seems to fail because of float rounding, I proposed a few two-char changes to hopefully fix that. I'm not sure if all of them are necessary, reject the ones that don't make sense to you.

If you want to add the constraint that the base distribution needs to be univariate, then after that I'm out of nits to pick and happy to approve.

EDIT: Oh and on the formatter thing, yeah, if the formatter makes it ugly I'm happy to still go with the what the formatter does, for consistency and ease.

mhauru commented on src/families/location_scale.jl:
```diff
@@ -30,8 +30,8 @@ This is necessary to guarantee stable convergence.
-    dist::ContinuousDistribution;
-    scale_eps::T=sqrt(eps(T)),
+    dist::ContinuousUnivariateDistribution;
+    scale_eps::T=eps(T)^(1//4),
```
I'm curious what provoked the `1//4` power.

Red-Portal replied:
This was an arbitrary choice because `1//2` seem to result in unstable behavior in some cases. I think I will just fix it to 1e-4 or 1e-3 regardless of the type. I don't think we expect posteriors with a variance of 1e-8 to exist in practice.

mhauru reviewed (approved):
Thanks @Red-Portal! Some Enzyme test now seems to fail, but I'll approve since the code itself looks good. Not immediately obvious to me what the issue is, feel free to look into it, or I can try to dig into it, hopefully sometime in the next couple of days.

Red-Portal commented:
@mhauru The enzyme [issue](https://github.com/EnzymeAD/Enzyme.jl/issues/1808) is due to an LLVM update. It should be resolved automatically pretty soon.

mhauru reviewed (approved):
Thanks @Red-Portal, looks great! I have no idea what the right value for `scale_eps` is, happy to take your word for it.
